### ISSUE TITLE
方便加微信交流

### ISSUE THREAD

**Issue by 1chenchen22** [PinJui/FDRL#2]
您好，我复现的在raf-db上是0.88的识别率吗，方便加微信交流一下这个程序吗chenxue19991020

**Comment by PinJui** [comment]
Hi, I'm sorry that I don't have Wechat. If you have any issues using this code, please feel free to write here.

**Comment by 1chenchen22** [comment]
> Hi, I'm sorry that I don't have Wechat. If you have any issues using this code, please feel free to write here.

Hello, I'm a beginner and I don't quite understand the class InterRM class. Specifically, I'm confused about the return self.delta * features + (1 - self.delta) * features_hat line.

In the InterRM class, the role of self.delta is to adjust the weights of the Intra-aware feature and Inter-aware feature, essentially representing the δ in the formula. Is my understanding correct?

The requirement in the text is to combine the j-th intra-aware feature and the j-th inter-aware feature to calculate the j-th importance-aware feature 
�
�
,
�
∈
�
�
y 
i,j
​
 ∈R 
D
  for the i-th facial image as follows:
�
�
,
�
=
�
�
�
,
�
+
(
1
−
�
)
�
^
�
,
�
 for 
�
=
1
,
2
,
…
,
�
y 
i,j
​
 =δf 
i,j
​
 +(1−δ) 
f
^
​
  
i,j
​
  for j=1,2,…,M
where δ represents the regularization parameter that balances the intra-aware feature and the inter-aware feature. Finally, a set of importance-aware features is added to obtain the final expression feature:
�
�
=
∑
�
=
1
�
�
�
,
�
y 
i
​
 =∑ 
j=1
M
​
 y 
i,j
​
 
where 
�
�
∈
�
�
y 
i
​
 ∈R 
D
  represents the expression feature for the i-th facial image.

In the code, return self.delta * features + (1 - self.delta) * features_hat, features is defined as a new empty list within the forward method of InterRM and does not have a direct relationship with IntraRM. In the line return self.delta * features + (1 - self.delta) * features_hat, features is the result processed by InterRM and does not have a direct correspondence with the operations in IntraRM. Therefore, it seems challenging to implement the description of formulas 11 and 12 from the original text.

Question 2: In my reproduction on the AffectNet dataset, there is oscillation in the accuracy of the test set, and there is overfitting. Have you encountered such a situation, and every time the training results are different? Can you provide guidance on how to address this issue?

**Comment by 1chenchen22** [comment]
Hello, I'm a beginner and I don't quite understand the class InterRM class. Specifically, I'm confused about the return self.delta * features + (1 - self.delta) * features_hat line.

In the InterRM class, the role of self.delta is to adjust the weights of the Intra-aware feature and Inter-aware feature, essentially representing the δ in the formula. Is my understanding correct?
![FZ2)J}`~X{%HZW758T(SI6P](https://github.com/PinJui/FDRL/assets/131838869/9b3436e7-a260-4549-9be6-710f190dcff2)
In the code, return self.delta * features + (1 - self.delta) * features_hat, features is defined as a new empty list within the forward method of InterRM and does not have a direct relationship with IntraRM. In the line return self.delta * features + (1 - self.delta) * features_hat, features is the result processed by InterRM and does not have a direct correspondence with the operations in IntraRM. Therefore, it seems challenging to implement the description of formulas 11 and 12 from the original text.

Question 2: In my reproduction on the AffectNet dataset, there is oscillation in the accuracy of the test set, and there is overfitting. Have you encountered such a situation, and every time the training results are different? Can you provide guidance on how to address this issue?
![QQ图片20240303143249](https://github.com/PinJui/FDRL/assets/131838869/be59e427-6ffc-4856-a7f1-6300310ade25)

**Comment by PinJui** [comment]
Thank you for your questions. The answers are as following:
1. Yes, `self.delta` in the InterRM module represents the *"...regularization parameter that balances the intra-aware feature and the inter-aware feature"* in the paper. 
2. Yes, I think you are correct. I mistakenly used the projected feature instead of the output directly from the IntraRM.
3. In my experience, I haven't encountered such situation. However, AffectNet is a quite imbalanced dataset and you may consider using some imbalance data sampler to address the overfitting issue.

Overall, I really appreciate for pointing out the error in this repo. I will make the correction as soon as possible. However, since I don't have those expression datasets anymore, I might not be able to reproduce the experiment results. Thank you!

**Comment by 1chenchen22** [comment]
> Thank you for your questions. The answers are as following:
> 
> 1. Yes, `self.delta` in the InterRM module represents the _"...regularization parameter that balances the intra-aware feature and the inter-aware feature"_ in the paper.
> 2. Yes, I think you are correct. I mistakenly used the projected feature instead of the output directly from the IntraRM.
> 3. In my experience, I haven't encountered such situation. However, AffectNet is a quite imbalanced dataset and you may consider using some imbalance data sampler to address the overfitting issue.
> 
> Overall, I really appreciate for pointing out the error in this repo. I will make the correction as soon as possible. However, since I don't have those expression datasets anymore, I might not be able to reproduce the experiment results. Thank you!

Thank you very much for your detailed answers! I appreciate that my questions helped you identify and correct the error in the repository. I understand that you may not be able to reproduce the experiment results anymore, but I still greatly appreciate your efforts.

I have made these modifications, and I'm not sure if they are correct. Among them, features_hat is the value returned by InterRM. I look forward to your feedback on the modifications.
![8G8PY A630BD4FP`B278N{E](https://github.com/PinJui/FDRL/assets/131838869/466a3343-32de-4503-bff2-0a86e88503b9)

 If you need any further assistance or feedback, please feel free to let me know. Once again, thank you for your time and patience!

**Comment by PinJui** [comment]
I've made the correction similar to your figure and kindly request testing if anyone possesses the RAF-DB dataset. Please feel free to ask any questions. If you find the problem resolved, you may close the issue. Thank you once again!

**Comment by 1chenchen22** [comment]
Hello, I trained my model on three different devices five times, each for 40 epochs, and obtained a suddenly low accuracy on the RAF-DB dataset. I've tried making changes to InterRM and keeping it unchanged, but the accuracy remains at 38.62%. Previously, the accuracy was much higher at 88%. I haven't modified the code or the dataset, and it's not an issue with the devices. What could be the reason for this sudden drop in accuracy?
![image](https://github.com/PinJui/FDRL/assets/131838869/e69d7714-95ef-45f3-ba5d-ecb56f00ebfb)

**Comment by PinJui** [comment]
I'm curious if this is related to my latest modification?

**Comment by 1chenchen22** [comment]
No, this has nothing to do with your latest revision, I tried to train with the previous version and the accuracy was very low,

**Comment by 1chenchen22** [comment]
Hello, I encountered an issue where the accuracy on the RAF dataset remains at 38%, not increasing, stuck in a local minimum, not converging. Could you please provide the previously trained .pth pre-trained weight file? Thank you for your help, sorry for the inconvenience.

**Comment by 1chenchen22** [comment]
Hello, I encountered an issue where the accuracy on the RAF dataset remains at 38%, not increasing, stuck in a local minimum, not converging. Could you please provide the previously trained .pth pre-trained weight file? Thank you for your help, sorry for the inconvenience.

**Comment by PinJui** [comment]
Hi, I'm sorry that I don't have the weight you request now and I think I don't have the data to reproduce the experiments. Maybe you can consider modify some training settings if it is related to the convergence issue.

**Comment by 1chenchen22** [comment]
Okay, thank you

**Comment by PinJui** [comment]
Since this issue is no longer active, I'll close it. However, feel free to reopen it if you have any further questions. Thank you very much!
